1307-x86-tpm.c-code-for-early-hashing-and-extending-PCR.patch

```diff
@@ -0,0 +1,530 @@
++    uint32_t hashDataSize;      /* 0-64, inclusive */
++    uint8_t hashData[];
++} __packed;
++
++struct sha1_complete_extend_rsp {
++    struct tpm_rsp_hdr h;
++    uint8_t hashValue[SHA1_DIGEST_SIZE];
++    uint8_t outDigest[SHA1_DIGEST_SIZE];
++} __packed;
++
++struct TPM12_PCREvent {
++    uint32_t PCRIndex;
++    uint32_t Type;
++    uint8_t Digest[SHA1_DIGEST_SIZE];
++    uint32_t Size;
++    uint8_t Data[];
++};
++
++struct txt_ev_log_container_12 {
++    char        Signature[20];      /* "TXT Event Container", null-terminated */
++    uint8_t     Reserved[12];
++    uint8_t     ContainerVerMajor;
++    uint8_t     ContainerVerMinor;
++    uint8_t     PCREventVerMajor;
++    uint8_t     PCREventVerMinor;
++    uint32_t    ContainerSize;      /* Allocated size */
++    uint32_t    PCREventsOffset;
++    uint32_t    NextEventOffset;
++    struct TPM12_PCREvent   PCREvents[];
++};
++
++#ifdef __EARLY_TPM__
++/*
++ * TPM1.2 is required to support commands of up to 1101 bytes, vendors rarely
++ * go above that. Limit maximum size of block of data to be hashed to 1024.
++ */
++#define MAX_HASH_BLOCK      1024
++#define CMD_RSP_BUF_SIZE    (sizeof(struct sha1_update_cmd) + MAX_HASH_BLOCK)
++
++union cmd_rsp {
++    struct sha1_start_cmd start_c;
++    struct sha1_start_rsp start_r;
++    struct sha1_update_cmd update_c;
++    struct sha1_update_rsp update_r;
++    struct sha1_complete_extend_cmd finish_c;
++    struct sha1_complete_extend_rsp finish_r;
++    uint8_t buf[CMD_RSP_BUF_SIZE];
++};
++
++static void tpm_hash_extend12(unsigned loc, uint8_t *buf, unsigned size,
++                              unsigned pcr, uint8_t *out_digest)
++{
++    union cmd_rsp cmd_rsp;
++    unsigned max_bytes = MAX_HASH_BLOCK;
++    unsigned o_size = sizeof(cmd_rsp);
++
++    request_locality(loc);
++
++    cmd_rsp.start_c = (struct sha1_start_cmd) {
++        .h.tag = swap16(TPM_TAG_RQU_COMMAND),
++        .h.paramSize = swap32(sizeof(struct sha1_start_cmd)),
++        .h.ordinal = swap32(TPM_ORD_SHA1Start),
++    };
++
++    send_cmd(loc, cmd_rsp.buf, sizeof(struct sha1_start_cmd), &o_size);
++
++    // assert (o_size >= sizeof(struct sha1_start_rsp));
++
++    if ( max_bytes > swap32(cmd_rsp.start_r.maxNumBytes) )
++        max_bytes = swap32(cmd_rsp.start_r.maxNumBytes);
++
++    while ( size > 64 ) {
++        if ( size < max_bytes )
++            max_bytes = size & ~(64 - 1);
++
++        o_size = sizeof(cmd_rsp);
++
++        cmd_rsp.update_c = (struct sha1_update_cmd){
++            .h.tag = swap16(TPM_TAG_RQU_COMMAND),
++            .h.paramSize = swap32(sizeof(struct sha1_update_cmd) + max_bytes),
++            .h.ordinal = swap32(TPM_ORD_SHA1Update),
++            .numBytes = swap32(max_bytes),
++        };
++        memcpy(cmd_rsp.update_c.hashData, buf, max_bytes);
++
++        send_cmd(loc, cmd_rsp.buf, sizeof(struct sha1_update_cmd) + max_bytes,
++                 &o_size);
++
++        // assert (o_size >= sizeof(struct sha1_update_rsp));
```

Comment:
I think it's documentation of an assumption made by the code. @krystian-hebel can confirm if that's the case.

Reply:
Honestly, I don't remember. I think it would be nice to check for those, but maybe not in early environment. It isn't linked with "regular" Xen code, so adding normal assertions would double the space used for messages.

Reply:
> I think it would be nice to check for those, but maybe not in early environment.

Well, this code is inside `#ifdef __EARLY_TPM__`.

Reply:
`if`-statement now: https://github.com/QubesOS/qubes-vmm-xen/pull/160/files#diff-cabb4c53c3369a702a10a9ade950a5e7c28e402a5ed35b50c96b50ea6563a32aR517-R518